Marlin/src/feature/easythreed_ui.cpp

```diff
@@ -0,0 +1,242 @@
+  if (blink_interval_ms >= LED_ON)  { WRITE(EASYTHREED_LED_PIN,  LOW); return; } // ON
+
+  const millis_t ms = millis();
+  if (prev_blink_interval_ms != blink_interval_ms) {
+    prev_blink_interval_ms = blink_interval_ms;
+    blink_start_ms = ms;
+  }
+  if (PENDING(ms, blink_start_ms + blink_interval_ms))
+    WRITE(EASYTHREED_LED_PIN, LOW);
+  else if (PENDING(ms, blink_start_ms + 2 * blink_interval_ms))
+    WRITE(EASYTHREED_LED_PIN, HIGH);
+  else
+    blink_start_ms = ms;
+}
+
+//
+// Filament Load/Unload Button
+//
+void EasythreedUI::loadButton() {
+  if (printingIsActive()) return;
+
+  enum FilamentStatus : uint8_t { FS_IDLE, FS_PRESS, FS_CHECK, FS_PROCEED };
+  static uint8_t filament_status = FS_IDLE;
+  static millis_t filament_time = 0;
+
+  switch (filament_status) {
+
+    case FS_IDLE:
+      if (!READ(BTN_RETRACT) || !READ(BTN_FEED)) {                  // If feed or retract button is pressed...
+        filament_status++;                                          // ...proceed to next test.
+        filament_time = millis();
+      }
+      break;
+
+    case FS_PRESS:
+      if (ELAPSED(millis(), filament_time + BTN_DEBOUNCE_MS)) {     // After a short debounce delay...
+        if (!READ(BTN_RETRACT) || !READ(BTN_FEED)) {                // ...if button(s) still pressed...
+          thermalManager.setTargetHotend(EXTRUDE_MINTEMP + 10, 0);  // Start heating up
+          blink_interval_ms = LED_BLINK_7;                          // Set the LED to blink fast
+          filament_status++;
+        }
+        else
+          filament_status = FS_IDLE;                                // Button(s) not pressed long enough
+      }
+      break;
+
+    case FS_CHECK:
+      if (READ(BTN_RETRACT) && READ(BTN_FEED)) {                    // Buttons are both released?
+        blink_interval_ms = LED_ON;                                 // LED on steady
+        filament_status = FS_IDLE;
+        thermalManager.disable_all_heaters();
+      }
+      else if (thermalManager.hotEnoughToExtrude(0)) {              // Is the hotend hot enough to move material?
+        filament_status++;                                          // Proceed to feed / retract.
+        blink_interval_ms = LED_BLINK_5;                            // Blink ~3 times per second
+      }
+      break;
+
+    case FS_PROCEED: {
+      // Feed or Retract just once. Hard abort all moves and return to idle on button release.
+      static bool flag = false;
+      if (READ(BTN_RETRACT) && READ(BTN_FEED)) {                    // Feed / retract button released?
+        flag = false;                                               // Restore flag to false
+        filament_status = FS_IDLE;                                  // Go back to idle state
+        quickstop_stepper();                                        // Hard-stop all the steppers ... now!
+        thermalManager.disable_all_heaters();                       // And disable all the heaters
```

Comment:
The feed / retract methodology employed here is … interesting. Also, it is somewhat likely to corrupt the `current_position`. There is a possibility that the machine may be in motion, or in the middle of `G29`, or otherwise not idle when this code is processed.